documents/system-design-document.md

```diff
@@ -0,0 +1,455 @@
+    '409':
+      description: Conflict - Email already exists
+
+DELETE /api/users/{user_id}:
+  summary: Deletes a specific user by their ID and all associated tasks
+  parameters:
+    - name: user_id
+      in: path
+      required: true
+      schema:
+        type: integer
+      description: The unique identifier of the user to delete
+  responses:
+    '204':
+      description: No Content - User and associated tasks successfully deleted
+    '404':
+      description: User not found
+    '400':
+      description: Bad request, possibly due to an invalid user ID
+
+POST /api/tasks:
+  summary: Creates a new task for the current user, optionally associating it with description and tags
+  requestBody:
+    content:
+      application/json:
+        schema:
+          type: object
+          properties:
+            name:
+              type: string
+              description: The name of the task
+            description:
+              type: string
+              description: A detailed description of the task
+            tags:
+              type: array
+              description: A list of tag IDs to associate with the task
+              items:
+                type: integer
+          required:
+            - name
+  responses:
+    '201':
+      description: Task successfully created
+      content:
+        application/json:
+          schema:
+            type: object
+            properties:
+              id:
+                type: integer
+                description: The unique identifier of the newly created task
+    '400':
+      description: Bad Request - Invalid input
+
+GET /api/tasks/:
+  summary: Returns a list of all tasks for the current user, optionally filtered by name, tag, or status
+  parameters:
+    - name: name
+      in: query
+      required: false
+      schema:
+        type: string
+      description: Filter tasks by name
+    - name: tag
+      in: query
+      required: false
+      schema:
+        type: string
+      description: Filter tasks by tag name
+    - name: status
+      in: query
+      required: false
+      schema:
+        type: string
+      description: Filter tasks by status
+  responses:
+    '200':
+      description: A JSON array of tasks matching the criteria or all tasks if no parameters are provided
+      content:
+        application/json:
+          schema: 
+            type: array
+            items:
+              type: object
+              properties:
+                id:
+                  type: integer
+                  description: Unique identifier for the task
+                name:
+                  type: string
+                  description: Name of the task
+                description:
+                  type: string
+                  description: Detailed description of the task
+                status:
+                  type: string
+                  description: Current status of the task (e.g., 'Created', 'Committed', 'In Progress', 'Completed')
```

Comment:
As commented above, if we create a separate `Task` object, the response payload would simply be array of `Task`.